File: LeetCode/Java/src/solutions/p1472_DesignBrowserHistory/P1472.java
package solutions.p1472_DesignBrowserHistory;

import java.util.ArrayList;
import java.util.List;

public class P1472 {
    class BrowserHistory {

        /*
         * Solution v1: Brute Force
         *
         * Runtime: 59 ms @ (beats) 54.85%
         * Memory Usage: 50.5 MB @ (beats) 53.18%
         */

        List<String> history;
        int curr;

        public BrowserHistory(String homepage) {
            this.history = new ArrayList<>();
            this.history.add(homepage);
            this.curr = 1;
        }

        public void visit(String url) {
            while (this.curr < this.history.size()) {
                this.history.remove(this.history.size() - 1);
            }
            this.history.add(url);
            this.curr = this.history.size();
        }

        public String back(int steps) {
            if (this.curr - steps < 0) {
                this.curr = 1;
            } else {
                this.curr -= steps;
            }
            return this.history.get(this.curr - 1);
        }

        public String forward(int steps) {
            if (this.curr + steps >= this.history.size()) {
                this.curr = this.history.size();
            } else {
                this.curr += steps;
            }
            return this.history.get(this.curr - 1);
        }
    }
}

File: LeetCode/Python/solutions/p1472.py
class BrowserHistory:

    #
    # Solution v1: Brute Force
    #
    # Runtime: 225 ms @ (beats) 75.17%
    # Memory Usage: 16.6 MB @ (beats) 85.99%
    #
    # def __init__(self, homepage: str):
    #     self.history = [homepage]
    #     self. curr = 0
    #
    # def visit(self, url: str) -> None:
    #     self.history = self.history[:self.curr + 1]
    #     self.history.append(url)
    #     self.curr = len(self.history) - 1
    #
    # def back(self, steps: int) -> str:
    #     self.curr -= steps
    #     if self.curr < 0:
    #         self.curr = 0
    #
    #     return self.history[self.curr]
    #
    # def forward(self, steps: int) -> str:
    #     self.curr += steps
    #     if self.curr >= len(self.history):
    #         self.curr = len(self.history) - 1
    #
    #     return self.history[self.curr]

    #
    # Solution v1.1: Brute Force
    # Improvement
    #
    # Runtime: 205 ms @ (beats) 98.55%
    # Memory Usage: 16.8 MB @ (beats) 37.33%
    #
    def __init__(self, homepage: str):
        self.history = [homepage]
        self.curr = 1

    def visit(self, url: str) -> None:
        self.history = self.history[:self.curr]
        self.history.append(url)
        self.curr = len(self.history)

    def back(self, steps: int) -> str:
        self.curr = max(1, self.curr - steps)
        return self.history[self.curr - 1]

    def forward(self, steps: int) -> str:
        self.curr = min(len(self.history), self.curr + steps)
        return self.history[self.curr - 1]
